### PR TITLE
virtcontainers: POC support for coldplugging PCI devices

### PR DESCRIPTION
**Pull request by eguzman3**
Adds support for coldplugging PCI devices at VM boot rather than hotplugging.
Adds two cli config options
`pci_coldplug_groups` defines a list of IOMMU groups that should be coldplugged when supplied via `--device=/dev/vfio/XX`
`pci_coldplug_device_opts` defines a list of per-device vfio-pci options to be passed on the qemu cli when coldplugging these devices

Fixes: #704

Signed-off-by: Edward Guzman <eguzman@nvidia.com>

### CODE REVIEW

**Comment by caoruidong** [comment]
Thank you for raising this! But if you want to modify govmm codes, you should also raise a PR in https://github.com/intel/govmm. When that PR gets merged, change this PR to 2 commits. The first one uses dep to import vendor changes, the second one is changes in this repo.

File: virtcontainers/pkg/oci/utils.go [hunk 1]
```diff
@@ -222,7 +222,81 @@ func newLinuxDeviceInfo(d spec.LinuxDevice) (*config.DeviceInfo, error) {
...
+	for _, g := range hConfig.PCIColdplugGroups {
+		if iommuGroup == strings.TrimSpace(g) {
+			return true
+		}
+	}
+	return false
+}
+
+// getBDF returns the BDF of pci device
+// Expected input strng format is [<domain>]:[<bus>][<slot>].[<func>] eg. 0000:02:10.0
```
Comment (jodh-intel) [inline virtcontainers/pkg/oci/utils.go]: Shouldn't that be (missing colon after `[<bus>]`?):

```
// Expected input strng format is [<domain>]:[<bus>]:[<slot>].[<func>] eg. 0000:02:10.0
```

File: virtcontainers/pkg/oci/utils.go [hunk 2]
```diff
@@ -222,7 +222,81 @@ func newLinuxDeviceInfo(d spec.LinuxDevice) (*config.DeviceInfo, error) {
...
+			return nil, err
+		}
+		bdfs = append(bdfs, deviceBDF)
+	}
+
+	return bdfs, nil
+}
+
+func checkForPCIColdPlug(device spec.LinuxDevice, hConfig *vc.HypervisorConfig) bool {
+	if !strings.Contains(device.Path, "/dev/vfio/") {
```
Comment (jodh-intel) [inline virtcontainers/pkg/oci/utils.go]: Should this be `strings.HasPrefix()`?

File: virtcontainers/pkg/oci/utils.go [hunk 3]
```diff
@@ -222,7 +222,81 @@ func newLinuxDeviceInfo(d spec.LinuxDevice) (*config.DeviceInfo, error) {
...
+		return false
+	}
+
+	pathTokens := strings.Split(device.Path, "/")
+	if len(pathTokens) != 4 {
```
Comment (jodh-intel) [inline virtcontainers/pkg/oci/utils.go]: Could you add a comment documenting the expected format so this token value is clearer?

File: cli/config/configuration.toml.in [hunk 4]
```diff
@@ -140,13 +140,27 @@ enable_iothreads = @DEFENABLEIOTHREADS@
...
+# A comma-separated list of IOMMU groups to coldplug into the VM.
+# The devices belong to these groups are only coldplugged if the
+# corresponding IOMMU group is passed using --device=/dev/vfio/XX
+# Example: "5, 7, 11"
+pci_coldplug_groups = "@DEFPCICOLDPLUGGROUPS@"
+
+# A list of parameters to pass to the vfio-pci driver when coldplugging
+# devices. This is a semicolon-separated list of device option lists where
+# a device option list is comma-separated list where the first entry is
+# the device's BDF. All devices that will be coldplugged should be added
```
Comment (jodh-intel) [inline cli/config/configuration.toml.in]: I'm finding this a little difficult to read. Could it be changed to something like this without losing the meaning?:

> A semi-colon separated list of parameters to pass to the vfio-pci driver when coldplugging
> devices. Each list element is itself a comma separated list of options which must start with the BDF of the device to be cold-plugged and optionally followed by one or more device-specific options.

File: virtcontainers/hypervisor.go [hunk 5]
```diff
@@ -241,6 +241,17 @@ type HypervisorConfig struct {
...
+	// PCIColdplugGroups is used to indicate which IOMMU groups' devices can be coldplugged
+	PCIColdplugGroups []string
+
+	// PCIColdplugGroups is used to indicate which PCI Devices should be coldplugged
+	// if they are passed with --device and their IOMMU group appears in PCIColdplugGroups
+	PCIColdplugDevices []string
+
+	// PCIColdplugDeviceOpts is used to indicate what paramaters
+	// should be passed to vfio-pci for coldplugged devices
+	PCIColdplugDeviceOpts []string
```
Comment (jodh-intel) [inline virtcontainers/hypervisor.go]: I'm not really clear why `PCIColdplugDeviceOpts` is separate from `PCIColdplugDeviceOpts`. How about something like:

```
type PCIDevice {
    Name      string
    Options []string
}

type HypervisorConfig struct {
        :

    PCIColdplugDevices []PCIDevice
}

```

Splitting the bdf and its options apart would also solve the issue in `checkForPCIColdPlug()` I think.

File: virtcontainers/pkg/oci/utils.go [hunk 6]
```diff
@@ -222,7 +222,81 @@ func newLinuxDeviceInfo(d spec.LinuxDevice) (*config.DeviceInfo, error) {
...
+	return false
+}
+
+// getBDF returns the BDF of pci device
+// Expected input strng format is [<domain>]:[<bus>][<slot>].[<func>] eg. 0000:02:10.0
+func getBDF(deviceSysStr string) (string, error) {
+	tokens := strings.Split(deviceSysStr, ":")
+
+	if len(tokens) != 3 {
+		return "", fmt.Errorf("Incorrect number of tokens found while parsing bdf for device : %s", deviceSysStr)
```
Comment (jodh-intel) [inline virtcontainers/pkg/oci/utils.go]: I'd consider making the `3` a `const` and adding that to the returned error so we log `3` vs. `len(tokens)`.

File: virtcontainers/pkg/oci/utils.go [hunk 7]
```diff
@@ -222,7 +222,81 @@ func newLinuxDeviceInfo(d spec.LinuxDevice) (*config.DeviceInfo, error) {
...
+	}
+
+	bdfs, err := getBDFsFromIOMMU(iommuGroup)
+	if err != nil {
+		return false
+	}
+
+	for _, bdf := range bdfs {
+		for _, opts := range hConfig.PCIColdplugDeviceOpts {
+			if strings.Contains(opts, bdf) {
```
Comment (jodh-intel) [inline virtcontainers/pkg/oci/utils.go]: What if a devices options contained a reference to another BDF? If that's possible, this won't be reliable.

**Comment by WeiZhang555** [comment]
Hi @eguzman3 , thanks for your PR! Can you address comment https://github.com/kata-containers/runtime/issues/704#issuecomment-419232969 ? We need a clear understanding about the scenario.

**Comment by raravena80** [comment]
Hi, @eguzman3 any updates on this? thx!

**Comment by jodh-intel** [comment]
Ping @eguzman3.

**Comment by flx42** [comment]
I will try to run some tests on a system we have that 16 GPUs with a large BAR size:
https://www.nvidia.com/en-us/data-center/dgx-2/
But I'm likely to encounter other issues unrelated to coldplug/hotplug, so if it's not conclusive we will have to table this PR for now.

**Comment by caoruidong** [comment]
@flx42 Do you solve other problems?

**Comment by raravena80** [comment]
@eguzman3 any updates? (from your weekly kata herder)

**Comment by jodh-intel** [comment]
Ok - closing. Let us know if you want this re-opened...

**Pull request closed without merging**